---
title: "curatoR-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{curatoR-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>")
```

The package curatoR contains functions that are aimed to help the analysis of genetic data. More specifically, it aids the analysis of single-cell RNA sequencing data. However, some functions could still be used for general use, such as for wrangling matrices. Most of the functions in this package originated from Rui's project at the Pavlidis Lab, which can be found in <https://github.com/xnrxng/PavlabSZProject/>. Some functions can be used by the curators at the Pavlidis lab.

This package is meant to be used for the programming language R. The version of R used to create this project was 4.4.1.

This document serves as a tutorial on how to use the functions. A more general overview of each function can also be seen through `?function_name`.

Once you have installed the package, load the library:

```{r setup}
library(curatoR)
```

## 1. load_w_new_col()

Loads a dataset from an URL and changes its column names.

This function takes 3 parameters:

-   dataset_path: URL to the dataset of interest as a string.
-   col_names: a vector of strings that are the column names.
-   delimiter: the dataset's delimiter as a string.

Here we show an example of how to use the function through an URL that leads to the mtcars dataset. More information on mtcars is available through `?mtcars`.

```{r}
dataset_url <- "https://raw.githubusercontent.com/plotly/datasets/master/mtcars.csv"

new_col_names <- c("model", "milespergallon", "cylinder_number", "displacement", "horsepower", "rear_axle_ratio", "weight", "quartermiletime", "engine", "transmission", "forwardgears", "carb")

comma_delimiter <- ","

loaded_dataset <- load_w_new_col(dataset_url, new_col_names, comma_delimiter)

# head() loads the first 6 rows of the tibble
head(loaded_dataset)
```

Note that the output is a tibble.

## 2. create_combined_matrix()

Authors usually provide each matrix in three files: barcodes.tsv, features.tsv, matrix.mtx. This function combines each matrix with its appropriate genes as rows and barcodes as columns.

This function takes 2 parameters:

-   specimenIDS: A character vector of specimen IDs. Each ID corresponds to a subdirectory or file prefix in the data directory.
-   initial_path: A string specifying the base path to the data directory containing subdirectories or files for each specimen. For example, if your files are in "data/data_raw/", that is your initial path.

In this case, my files are named "Patient1barcodes.tsv", "Patient1features.tsv" and "Patient1matrix.mtx". They are found under my "tests/testthat/files_helpers/" folder. Thus, this would be my initial_path.

```{r}
initial_path <- paste0(system.file("extdata/", package = "curatoR"), "/")
specimenIDs <- c("Patient1", "Patient2")
combined_matrices <- create_combined_matrix(specimenIDs, initial_path)

print(combined_matrices$Patient1)
print(combined_matrices$Patient2)
```

Note that the final output is a list of matrices.

## 3. generate_cell_types()

When authors provide matrices without unique cell IDs, this function creates unique cell IDs for each cell as well as sorts the matrices by cell type.

This function takes 3 parameters:

-   cell_types: Vector of cell types to process (e.g., c("T_cells", "B_cells")).
-   sample_list: List of matrices where rows are genes and columns are cells. Each matrix should have column names indicating cell types. Genes should be the same (same number and order) for all matrices.
-   prefix: Prefix to include in the column names of the output.

```{r}
sample1 <- matrix(c(1, 2, 3, 4, 5, 6), nrow = 3,
dimnames = list(c("Gene1", "Gene2", "Gene3"), c("T_cells", "B_cells")))

sample2 <- matrix(c(7, 8, 9, 10, 11, 12), nrow = 3,
dimnames = list(c("Gene1", "Gene2", "Gene3"), c("T_cells", "B_cells")))

sample_list <- list(sample1, sample2)

cell_types <- c("T_cells", "B_cells")

prefix <- "Sample_"

ct_list <- generate_cell_types(cell_types, sample_list, prefix)

print(ct_list$T_cells)
print(ct_list$B_cells)
```

Note that the output is a list of matrices, where each matrix corresponds to a cell type.

## 4. generate_cell_types_w_barcodes()

When you want to categorize matrices based on cell types, and the barcodes are linked to the cell types, you can use this function.

This function takes 3 parameters:

-   cell_types: Vector of cell types to process (e.g., c("T_cells", "B_cells")).
-   sample_list: List of matrices where rows are genes and columns are cells. Each matrix should have column names indicating cell types. Genes should be the same (same number and order) for all matrices.
-   cell_IDs: A dataframe that should have two columns: "cell_type" and "cell_ID".

```{r}
cell_IDs <- data.frame(
cell_type = c("T_cells", "B_cells", "T_cells", "B_cells",
"T_cells", "T_cells", "B_cells", "B_cells"),
cell_ID = c("Cell1", "Cell2", "Cell3", "Cell4",
"Cell5", "Cell6", "Cell7", "Cell8"))

sample1 <- matrix(
c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12),
nrow = 3,
dimnames = list(c("Gene1", "Gene2", "Gene3"),
c("Cell1", "Cell2", "Cell3", "Cell4")))

sample2 <- matrix(
c(13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24),
nrow = 3,
dimnames = list(c("Gene1", "Gene2", "Gene3"),
c("Cell5", "Cell6", "Cell7", "Cell8")))

sample_list <- list(sample1, sample2)

cell_types <- c("T_cells", "B_cells")

barcodes_list <- generate_cell_types_w_barcodes(cell_types, sample_list, cell_IDs)

print(barcodes_list$T_cells)
print(barcodes_list$B_cells)
```

Note that the output is a list of sparse matrices (one for each cell type in cell_types), with cell IDs as column names and shared row names.

## 5. cleanCtmat()

This function filters a matrix based on expression levels.

This function takes 3 parameters:

-   ctmat: A numeric matrix where rows represent genes and columns represent samples (e.g., cells).
-   geneThr: A numeric value between 0 and 1 specifying the minimum proportion of cells in which a gene must be expressed to be retained. The default is 0.02.
-   sampleThr: A numeric value between 0 and 1 specifying the minimum rank percentile for the number of genes expressed in a sample to retain that sample. The default is 0.05.

```{r}
set.seed(123)
ctmat <- matrix(sample(c(0, 1, 2, 3), 30, replace = TRUE),
nrow = 5, ncol = 6,
dimnames = list(
paste0("Gene", 1:5),
paste0("Cell", 1:6)))

cleaned_mm <- cleanCtmat(ctmat)

print(cleaned_mm)
```

It returns the filtered matrix.

## 6. do_cpm_log()

This function CPM normalizes a matrix. It also optionally log transforms it.

This function takes 2 parameters:

-   mtx: A matrix.
-   log: TRUE or FALSE. Whether to log-transform the matrix as well. Default is false.

```{r}
set.seed(123)
ctmat <- matrix(sample(c(0, 1, 2, 3), 30, replace = TRUE),
nrow = 5, ncol = 6,
dimnames = list(
paste0("Gene", 1:5),
paste0("Cell", 1:6)))

cpmlogged <- do_cpm_log(ctmat, log = TRUE)

print(cpmlogged)
```

It returns the transformed matrix.

## 7. create_pseudobulk()

This function creates a pseudobulk for an expression matrix.

This function takes 2 parameters:

-   expr: Expression matrix where columns are cell IDs and rows are gene names.
-   meta: Dataframe where each row should represent a cell. Cell IDs should be the rownames. It needs to have the columns "patientID", "disorder", "sex", and "age". "disorder" should have the values "yes" or "no".

```{r}
set.seed(123)
expr <- matrix(sample(0:10, 50, replace = TRUE), 
               nrow = 5, ncol = 10, 
               dimnames = list(paste0("Gene", 1:5), paste0("Cell", 1:10)))

meta <- data.frame(
  patientID = c("P1", "P2", "P3", "P1", "P1", "P2", "P2", "P3", "P3", "P1"),
  disorder = c("yes", "no", "no", "yes", "yes", "no", "no", "no", "no", "yes"),
  sex = c("male", "female", "female", "male", "male", "female", "female", "female", "female", "male"),
  age = c(70, 68, 64, 70, 70, 68, 68, 64, 64, 70),
  row.names = paste0("Cell", 1:10)
)

pseudo_bulk <- create_pseudo_bulk(expr, meta)

print(pseudo_bulk$meta)
print(pseudo_bulk$expr)
```

It returns a list composed of the pseudobulk as well as its associated metadata.

## 8. perform_DGE()

This function TMM normalizes a pseudobulk matrix and then performs DEA with the edgeR package.

This function takes 2 parameters:

-   PB: The output of create_pseudobulk.
-   design: To specify the groups of interest. E.g. design <- model.matrix(~ group, data = PB$meta)

```{r}
design <- model.matrix(~ group, data = pseudo_bulk$meta)

res <- perform_DGE(pseudo_bulk, design)

print(res)
```

It returns the DEA results.

## 9. limma_dge()

This function CPMLog normalizes a pseudobulk matrix and then performs DEA with the limma package.

This function takes 2 parameters:

-   PB: The output of create_pseudobulk.
-   design: To specify the groups of interest. E.g. design <- model.matrix(~ group, data = PB$meta)

```{r}
design <- model.matrix(~ group, data = pseudo_bulk$meta)

res_limma <- limma_dge(design, pseudo_bulk)

print(res_limma)
```
