<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>curatoR tutorial • curatoR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Montserrat-0.4.10/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="curatoR tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">curatoR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/curatoR-vignette.html">curatoR tutorial</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/xnrxng/curatoR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>curatoR tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/xnrxng/curatoR/blob/master/vignettes/curatoR-vignette.Rmd" class="external-link"><code>vignettes/curatoR-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>curatoR-vignette.Rmd</code></div>
    </div>

    
    
<p>The package curatoR contains functions that are aimed to help the
analysis of genetic data. More specifically, it aids the analysis of
single-cell RNA sequencing data. However, some functions could still be
used for general use, such as for wrangling matrices. Most of the
functions in this package originated from Rui’s project at the Pavlidis
Lab, which can be found in <a href="https://github.com/xnrxng/PavlabSZProject/" class="external-link uri">https://github.com/xnrxng/PavlabSZProject/</a>. Some
functions can be used by the curators at the Pavlidis lab.</p>
<p>This package is meant to be used for the programming language R. The
version of R used to create this project was 4.4.1.</p>
<p>This document serves as a tutorial on how to use the functions. A
more general overview of each function can also be seen through
<code>?function_name</code>.</p>
<p>Once you have installed the package, load the library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xnrxng/curatoR" class="external-link">curatoR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="load_w_new_col">1. load_w_new_col()<a class="anchor" aria-label="anchor" href="#load_w_new_col"></a>
</h2>
<p>Loads a dataset from an URL and changes its column names.</p>
<p>This function takes 3 parameters:</p>
<ul>
<li>dataset_path: URL to the dataset of interest as a string.</li>
<li>col_names: a vector of strings that are the column names.</li>
<li>delimiter: the dataset’s delimiter as a string.</li>
</ul>
<p>Here we show an example of how to use the function through an URL
that leads to the mtcars dataset. More information on mtcars is
available through <code><a href="https://rdrr.io/r/datasets/mtcars.html" class="external-link">?mtcars</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/plotly/datasets/master/mtcars.csv"</span></span>
<span></span>
<span><span class="va">new_col_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"model"</span>, <span class="st">"milespergallon"</span>, <span class="st">"cylinder_number"</span>, <span class="st">"displacement"</span>, <span class="st">"horsepower"</span>, <span class="st">"rear_axle_ratio"</span>, <span class="st">"weight"</span>, <span class="st">"quartermiletime"</span>, <span class="st">"engine"</span>, <span class="st">"transmission"</span>, <span class="st">"forwardgears"</span>, <span class="st">"carb"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">comma_delimiter</span> <span class="op">&lt;-</span> <span class="st">","</span></span>
<span></span>
<span><span class="va">loaded_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_w_new_col.html">load_w_new_col</a></span><span class="op">(</span><span class="va">dataset_url</span>, <span class="va">new_col_names</span>, <span class="va">comma_delimiter</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">32</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">12</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span>  (1): manufacturer</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (11): mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span></span>
<span><span class="co"># head() loads the first 6 rows of the tibble</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">loaded_dataset</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 12</span></span></span>
<span><span class="co">#&gt;   model   milespergallon cylinder_number displacement horsepower rear_axle_ratio</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Mazda …           21                 6          160        110            3.9 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Mazda …           21                 6          160        110            3.9 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Datsun…           22.8               4          108         93            3.85</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Hornet…           21.4               6          258        110            3.08</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Hornet…           18.7               8          360        175            3.15</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Valiant           18.1               6          225        105            2.76</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6 more variables: weight &lt;dbl&gt;, quartermiletime &lt;dbl&gt;, engine &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   transmission &lt;dbl&gt;, forwardgears &lt;dbl&gt;, carb &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Note that the output is a tibble.</p>
</div>
<div class="section level2">
<h2 id="create_combined_matrix">2. create_combined_matrix()<a class="anchor" aria-label="anchor" href="#create_combined_matrix"></a>
</h2>
<p>Authors usually provide each matrix in three files: barcodes.tsv,
features.tsv, matrix.mtx. This function combines each matrix with its
appropriate genes as rows and barcodes as columns.</p>
<p>This function takes 2 parameters:</p>
<ul>
<li>specimenIDS: A character vector of specimen IDs. Each ID corresponds
to a subdirectory or file prefix in the data directory.</li>
<li>initial_path: A string specifying the base path to the data
directory containing subdirectories or files for each specimen. For
example, if your files are in “data/data_raw/”, that is your initial
path.</li>
</ul>
<p>In this case, my files are named “Patient1barcodes.tsv”,
“Patient1features.tsv” and “Patient1matrix.mtx”. They are found under my
“tests/testthat/files_helpers/” folder. Thus, this would be my
initial_path.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initial_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/"</span>, package <span class="op">=</span> <span class="st">"curatoR"</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">specimenIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient1"</span>, <span class="st">"Patient2"</span><span class="op">)</span></span>
<span><span class="va">combined_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_combined_matrix.html">create_combined_matrix</a></span><span class="op">(</span><span class="va">specimenIDs</span>, <span class="va">initial_path</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">combined_matrices</span><span class="op">$</span><span class="va">Patient1</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 3 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;       Barcode1 Barcode2 Barcode3</span></span>
<span><span class="co">#&gt; gene1        1        .        .</span></span>
<span><span class="co">#&gt; gene2        .        2        4</span></span>
<span><span class="co">#&gt; gene3        3        .        5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">combined_matrices</span><span class="op">$</span><span class="va">Patient2</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 3 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">#&gt;       Barcode1 Barcode2 Barcode3</span></span>
<span><span class="co">#&gt; gene1        1        .        .</span></span>
<span><span class="co">#&gt; gene2        .        2        4</span></span>
<span><span class="co">#&gt; gene3        3        .        5</span></span></code></pre></div>
<p>Note that the final output is a list of matrices.</p>
</div>
<div class="section level2">
<h2 id="generate_cell_types">3. generate_cell_types()<a class="anchor" aria-label="anchor" href="#generate_cell_types"></a>
</h2>
<p>When authors provide matrices without unique cell IDs, this function
creates unique cell IDs for each cell as well as sorts the matrices by
cell type.</p>
<p>This function takes 3 parameters:</p>
<ul>
<li>cell_types: Vector of cell types to process (e.g., c(“T_cells”,
“B_cells”)).</li>
<li>sample_list: List of matrices where rows are genes and columns are
cells. Each matrix should have column names indicating cell types. Genes
should be the same (same number and order) for all matrices.</li>
<li>prefix: Prefix to include in the column names of the output.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="st">"B_cells"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">9</span>, <span class="fl">10</span>, <span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="st">"B_cells"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">sample1</span>, <span class="va">sample2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="st">"B_cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">prefix</span> <span class="op">&lt;-</span> <span class="st">"Sample_"</span></span>
<span></span>
<span><span class="va">ct_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_cell_types.html">generate_cell_types</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">sample_list</span>, <span class="va">prefix</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ct_list</span><span class="op">$</span><span class="va">T_cells</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 2 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;       Sample_T_cells_Sample1_1 Sample_T_cells_Sample2_2</span></span>
<span><span class="co">#&gt; Gene1                        1                        7</span></span>
<span><span class="co">#&gt; Gene2                        2                        8</span></span>
<span><span class="co">#&gt; Gene3                        3                        9</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ct_list</span><span class="op">$</span><span class="va">B_cells</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 2 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;       Sample_B_cells_Sample1_1 Sample_B_cells_Sample2_2</span></span>
<span><span class="co">#&gt; Gene1                        4                       10</span></span>
<span><span class="co">#&gt; Gene2                        5                       11</span></span>
<span><span class="co">#&gt; Gene3                        6                       12</span></span></code></pre></div>
<p>Note that the output is a list of matrices, where each matrix
corresponds to a cell type.</p>
</div>
<div class="section level2">
<h2 id="generate_cell_types_w_barcodes">4. generate_cell_types_w_barcodes()<a class="anchor" aria-label="anchor" href="#generate_cell_types_w_barcodes"></a>
</h2>
<p>When you want to categorize matrices based on cell types, and the
barcodes are linked to the cell types, you can use this function.</p>
<p>This function takes 3 parameters:</p>
<ul>
<li>cell_types: Vector of cell types to process (e.g., c(“T_cells”,
“B_cells”)).</li>
<li>sample_list: List of matrices where rows are genes and columns are
cells. Each matrix should have column names indicating cell types. Genes
should be the same (same number and order) for all matrices.</li>
<li>cell_IDs: A dataframe that should have two columns: “cell_type” and
“cell_ID”.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_IDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="st">"B_cells"</span>, <span class="st">"T_cells"</span>, <span class="st">"B_cells"</span>,</span>
<span><span class="st">"T_cells"</span>, <span class="st">"T_cells"</span>, <span class="st">"B_cells"</span>, <span class="st">"B_cells"</span><span class="op">)</span>,</span>
<span>cell_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell1"</span>, <span class="st">"Cell2"</span>, <span class="st">"Cell3"</span>, <span class="st">"Cell4"</span>,</span>
<span><span class="st">"Cell5"</span>, <span class="st">"Cell6"</span>, <span class="st">"Cell7"</span>, <span class="st">"Cell8"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">9</span>, <span class="fl">10</span>, <span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span>,</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell1"</span>, <span class="st">"Cell2"</span>, <span class="st">"Cell3"</span>, <span class="st">"Cell4"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">13</span>, <span class="fl">14</span>, <span class="fl">15</span>, <span class="fl">16</span>, <span class="fl">17</span>, <span class="fl">18</span>, <span class="fl">19</span>, <span class="fl">20</span>, <span class="fl">21</span>, <span class="fl">22</span>, <span class="fl">23</span>, <span class="fl">24</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">3</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span>,</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell5"</span>, <span class="st">"Cell6"</span>, <span class="st">"Cell7"</span>, <span class="st">"Cell8"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">sample1</span>, <span class="va">sample2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="st">"B_cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">barcodes_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_cell_types_w_barcodes.html">generate_cell_types_w_barcodes</a></span><span class="op">(</span><span class="va">cell_types</span>, <span class="va">sample_list</span>, <span class="va">cell_IDs</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">barcodes_list</span><span class="op">$</span><span class="va">T_cells</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;       Cell1 Cell3 Cell5 Cell6</span></span>
<span><span class="co">#&gt; Gene1     1     7    13    16</span></span>
<span><span class="co">#&gt; Gene2     2     8    14    17</span></span>
<span><span class="co">#&gt; Gene3     3     9    15    18</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">barcodes_list</span><span class="op">$</span><span class="va">B_cells</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 x 4 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;       Cell2 Cell4 Cell7 Cell8</span></span>
<span><span class="co">#&gt; Gene1     4    10    19    22</span></span>
<span><span class="co">#&gt; Gene2     5    11    20    23</span></span>
<span><span class="co">#&gt; Gene3     6    12    21    24</span></span></code></pre></div>
<p>Note that the output is a list of sparse matrices (one for each cell
type in cell_types), with cell IDs as column names and shared row
names.</p>
</div>
<div class="section level2">
<h2 id="cleanctmat">5. cleanCtmat()<a class="anchor" aria-label="anchor" href="#cleanctmat"></a>
</h2>
<p>This function filters a matrix based on expression levels.</p>
<p>This function takes 3 parameters:</p>
<ul>
<li>ctmat: A numeric matrix where rows represent genes and columns
represent samples (e.g., cells).</li>
<li>geneThr: A numeric value between 0 and 1 specifying the minimum
proportion of cells in which a gene must be expressed to be retained.
The default is 0.02.</li>
<li>sampleThr: A numeric value between 0 and 1 specifying the minimum
rank percentile for the number of genes expressed in a sample to retain
that sample. The default is 0.05.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">ctmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">30</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">5</span>, ncol <span class="op">=</span> <span class="fl">6</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cleanCtmat.html">cleanCtmat</a></span><span class="op">(</span><span class="va">ctmat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cleaned_mm</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Cell1 Cell2 Cell3 Cell4 Cell6</span></span>
<span><span class="co">#&gt; Gene1     2     1     3     2     2</span></span>
<span><span class="co">#&gt; Gene2     2     1     1     3     3</span></span>
<span><span class="co">#&gt; Gene3     2     1     1     0     1</span></span>
<span><span class="co">#&gt; Gene4     1     2     0     2     2</span></span>
<span><span class="co">#&gt; Gene5     2     0     1     2     1</span></span></code></pre></div>
<p>It returns the filtered matrix.</p>
</div>
<div class="section level2">
<h2 id="do_cpm_log">6. do_cpm_log()<a class="anchor" aria-label="anchor" href="#do_cpm_log"></a>
</h2>
<p>This function CPM normalizes a matrix. It also optionally log
transforms it.</p>
<p>This function takes 2 parameters:</p>
<ul>
<li>mtx: A matrix.</li>
<li>log: TRUE or FALSE. Whether to log-transform the matrix as well.
Default is false.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">ctmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fl">30</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">5</span>, ncol <span class="op">=</span> <span class="fl">6</span>,</span>
<span>dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,</span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cpmlogged</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do_cpm_log.html">do_cpm_log</a></span><span class="op">(</span><span class="va">ctmat</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cpmlogged</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Cell1    Cell2    Cell3    Cell4    Cell5    Cell6</span></span>
<span><span class="co">#&gt; Gene1 12.31144 12.20608 13.12237 12.31144  0.00000 12.31144</span></span>
<span><span class="co">#&gt; Gene2 12.31144 12.20608 12.02376 12.71690 13.81551 12.71690</span></span>
<span><span class="co">#&gt; Gene3 12.31144 12.20608 12.02376  0.00000  0.00000 11.61829</span></span>
<span><span class="co">#&gt; Gene4 11.61829 12.89922  0.00000 12.31144  0.00000 12.31144</span></span>
<span><span class="co">#&gt; Gene5 12.31144  0.00000 12.02376 12.31144  0.00000 11.61829</span></span></code></pre></div>
<p>It returns the transformed matrix.</p>
</div>
<div class="section level2">
<h2 id="create_pseudobulk">7. create_pseudobulk()<a class="anchor" aria-label="anchor" href="#create_pseudobulk"></a>
</h2>
<p>This function creates a pseudobulk for an expression matrix.</p>
<p>This function takes 2 parameters:</p>
<ul>
<li>expr: Expression matrix where columns are cell IDs and rows are gene
names.</li>
<li>meta: Dataframe where each row should represent a cell. Cell IDs
should be the rownames. It needs to have the columns “patientID”,
“disorder”, “sex”, and “age”. “disorder” should have the values “yes” or
“no”.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>               nrow <span class="op">=</span> <span class="fl">5</span>, ncol <span class="op">=</span> <span class="fl">10</span>, </span>
<span>               dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  patientID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span>, <span class="st">"P1"</span>, <span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span>, <span class="st">"P3"</span>, <span class="st">"P1"</span><span class="op">)</span>,</span>
<span>  disorder <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yes"</span>, <span class="st">"no"</span>, <span class="st">"no"</span>, <span class="st">"yes"</span>, <span class="st">"yes"</span>, <span class="st">"no"</span>, <span class="st">"no"</span>, <span class="st">"no"</span>, <span class="st">"no"</span>, <span class="st">"yes"</span><span class="op">)</span>,</span>
<span>  sex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"male"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"male"</span>, <span class="st">"male"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"female"</span>, <span class="st">"male"</span><span class="op">)</span>,</span>
<span>  age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">70</span>, <span class="fl">68</span>, <span class="fl">64</span>, <span class="fl">70</span>, <span class="fl">70</span>, <span class="fl">68</span>, <span class="fl">68</span>, <span class="fl">64</span>, <span class="fl">64</span>, <span class="fl">70</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseudo_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_pseudo_bulk.html">create_pseudo_bulk</a></span><span class="op">(</span><span class="va">expr</span>, <span class="va">meta</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pseudo_bulk</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span>
<span><span class="co">#&gt;              group_sample       group patientID disorder    sex age</span></span>
<span><span class="co">#&gt; 1  patientIDP2:disorderno  disorderno        P2       no female  68</span></span>
<span><span class="co">#&gt; 2  patientIDP3:disorderno  disorderno        P3       no female  64</span></span>
<span><span class="co">#&gt; 3 patientIDP1:disorderyes disorderyes        P1      yes   male  70</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pseudo_bulk</span><span class="op">$</span><span class="va">expr</span><span class="op">)</span></span>
<span><span class="co">#&gt;       patientIDP2:disorderno patientIDP3:disorderno patientIDP1:disorderyes</span></span>
<span><span class="co">#&gt; Gene1                     22                     25                      26</span></span>
<span><span class="co">#&gt; Gene2                     10                     25                      17</span></span>
<span><span class="co">#&gt; Gene3                     21                     16                      26</span></span>
<span><span class="co">#&gt; Gene4                     19                      9                      19</span></span>
<span><span class="co">#&gt; Gene5                     21                     18                      24</span></span></code></pre></div>
<p>It returns a list composed of the pseudobulk as well as its
associated metadata.</p>
</div>
<div class="section level2">
<h2 id="perform_dge">8. perform_DGE()<a class="anchor" aria-label="anchor" href="#perform_dge"></a>
</h2>
<p>This function TMM normalizes a pseudobulk matrix and then performs
DEA with the edgeR package.</p>
<p>This function takes 2 parameters:</p>
<ul>
<li>PB: The output of create_pseudobulk.</li>
<li>design: To specify the groups of interest. E.g. design &lt;-
model.matrix(~ group, data = PB$meta)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">pseudo_bulk</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform_DGE.html">perform_DGE</a></span><span class="op">(</span><span class="va">pseudo_bulk</span>, <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt;    gene         logFC   logCPM           LR    PValue       FDR</span></span>
<span><span class="co">#&gt; 1 Gene2 -0.3563918390 17.52844 2.104152e-01 0.6464422 0.9993739</span></span>
<span><span class="co">#&gt; 2 Gene1 -0.1532864102 17.96376 1.923554e-01 0.6609634 0.9993739</span></span>
<span><span class="co">#&gt; 3 Gene3  0.1917995995 17.75488 1.617865e-01 0.6875171 0.9993739</span></span>
<span><span class="co">#&gt; 4 Gene4  0.1513627368 17.37293 3.323124e-02 0.8553517 0.9993739</span></span>
<span><span class="co">#&gt; 5 Gene5  0.0003725106 17.76122 6.157728e-07 0.9993739 0.9993739</span></span>
<span><span class="co">#&gt;               test</span></span>
<span><span class="co">#&gt; 1 pseudobulk_edgeR</span></span>
<span><span class="co">#&gt; 2 pseudobulk_edgeR</span></span>
<span><span class="co">#&gt; 3 pseudobulk_edgeR</span></span>
<span><span class="co">#&gt; 4 pseudobulk_edgeR</span></span>
<span><span class="co">#&gt; 5 pseudobulk_edgeR</span></span></code></pre></div>
<p>It returns the DEA results.</p>
</div>
<div class="section level2">
<h2 id="limma_dge">9. limma_dge()<a class="anchor" aria-label="anchor" href="#limma_dge"></a>
</h2>
<p>This function CPMLog normalizes a pseudobulk matrix and then performs
DEA with the limma package.</p>
<p>This function takes 2 parameters:</p>
<ul>
<li>PB: The output of create_pseudobulk.</li>
<li>design: To specify the groups of interest. E.g. design &lt;-
model.matrix(~ group, data = PB$meta)</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">pseudo_bulk</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/limma_dge.html">limma_dge</a></span><span class="op">(</span><span class="va">design</span>, <span class="va">pseudo_bulk</span><span class="op">)</span></span>
<span><span class="co">#&gt; Removing intercept from test coefficients</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res_limma</span><span class="op">)</span></span>
<span><span class="co">#&gt;             logFC  AveExpr          t   P.Value adj.P.Val         B</span></span>
<span><span class="co">#&gt; Gene2 -0.16867848 17.35312 -0.2101962 0.8418116 0.9064321 -4.602487</span></span>
<span><span class="co">#&gt; Gene4  0.25314724 17.22803  0.4469408 0.6736126 0.9064321 -4.607448</span></span>
<span><span class="co">#&gt; Gene3  0.22700091 17.68798  0.6859200 0.5232677 0.9064321 -4.619125</span></span>
<span><span class="co">#&gt; Gene5  0.03126067 17.70526  0.1236196 0.9064321 0.9064321 -4.666733</span></span>
<span><span class="co">#&gt; Gene1 -0.11980888 17.91918 -0.9105933 0.4042704 0.9064321 -4.671289</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get_cell_type_file">10. get_cell_type_file()<a class="anchor" aria-label="anchor" href="#get_cell_type_file"></a>
</h2>
<p>This function takes in a raw file and wrangles it so that it is
according to how Gemma needs them. This function takes in 5
parameters:</p>
<ul>
<li>file_path: a string that contains the path to the raw file.</li>
<li>remove_prefix: whether to remove the barcodes’ prefix. Default is
FALSE.</li>
<li>remove_suffix: whether to remove the barcodes’ suffix. Defaulst is
FALSE.</li>
<li>id_col: a string. You can specify which column is the barcodes
column if there are more than one. Default is NULL.</li>
<li>type_col: a string. You can specify which column is the cell type
column if there are more than one. Default is NULL.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rui Xiang Yu, Nairuz Elazzabi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
